src/enclave/enclave_oe.c

```diff
@@ -311,38 +285,87 @@ static void _read_eeid_config()
     sgxlkl_enclave_state.config = cfg;
 }
 
+void sgxlkl_ensure_inside(const void* ptr, size_t sz)
+{
+    if (ptr != NULL && !oe_is_within_enclave(ptr, sz))
+        sgxlkl_fail("Unexpected: buffer is not in enclave memory.\n");
+}
+
+void sgxlkl_ensure_outside(const void* ptr, size_t sz)
+{
+    if (ptr != NULL && !oe_is_outside_enclave(ptr, sz))
+        sgxlkl_fail("Unexpected: buffer is in enclave memory.\n");
+}
+
 static void _copy_shared_memory(const sgxlkl_shared_memory_t* host)
 {
     const sgxlkl_enclave_config_t* cfg = sgxlkl_enclave_state.config;
 
-    /* Deep copy where necessary */
+    /* The host shared memory struct has been copied by OE (but not its
+     * contents). */
+    sgxlkl_ensure_inside(host, sizeof(sgxlkl_shared_memory_t));
 
-    sgxlkl_shared_memory_t* enc = &sgxlkl_enclave_state.shared_memory;
-    memset(enc, 0, sizeof(sgxlkl_shared_memory_t));
+    /* Temporary, volatile copy to make sure checks aren't reordered */
```

Comment:
`host` points to a `sgxlkl_shared_memory_t` structure copied into the enclave memory by OE. It's safe to access the structure inside the enclave directly, without worring about TOCTOU. Once all the checks pass, the structure pointed by `host` can be copied to sgxlkl_enclave_state.shared_memory directly, without using the volatile copy.